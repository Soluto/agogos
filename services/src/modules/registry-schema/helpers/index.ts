import proceedPolicy from './policy-helper';

export { proceedPolicy };
