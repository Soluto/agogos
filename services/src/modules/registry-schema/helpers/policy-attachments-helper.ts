import * as _ from 'lodash';
import * as opaHelper from '../../opa-helper';
import logger from '../../logger';
import { PolicyType, ResourceMetadata, ResourceRepository } from '../../resource-repository';
import resourceRepository from '../repository';
import { ResourceMetadataInput, PolicyInput } from '../types';

type TempLocalPolicyAttachment = { metadata: ResourceMetadataInput; path: string; type: PolicyType };

const policyMetadataComparer = (p1: PolicyInput, p2: PolicyInput) =>
  p1.metadata.namespace === p2.metadata.namespace && p1.metadata.name === p2.metadata.name;

export default class PolicyAttachmentsHelper {
  protected policyAttachmentsToSave: TempLocalPolicyAttachment[] = [];
  protected policyAttachmentsToDelete: PolicyInput[] = [];

  async sync(oldState: PolicyInput[], newState: PolicyInput[]) {
    // Policy definitions generated by plugins will always be re-generated here. This is probably
    // better than reading the previous gateway resource group just for this comparison
    const addedOrUpdatedPolicies = _.differenceWith(newState, oldState, _.isEqual);
    await this.generate(addedOrUpdatedPolicies);

    const removedPolicies = _.differenceWith(oldState, newState, policyMetadataComparer);
    this.policyAttachmentsToDelete.push(...removedPolicies);
  }

  private async generate(policies: PolicyInput[]) {
    for (const policy of policies) {
      if (!policyAttachmentStrategies[policy.type]) continue;

      const attachment = await policyAttachmentStrategies[policy.type].generate(policy);
      this.policyAttachmentsToSave.push(attachment);
    }
  }

  async writeToRepo() {
    for (const attachment of this.policyAttachmentsToSave) {
      await policyAttachmentStrategies[attachment.type].saveToRepo(resourceRepository, attachment);
    }

    for (const policy of this.policyAttachmentsToDelete) {
      await policyAttachmentStrategies[policy.type].removeFromRepo(resourceRepository, policy.metadata);
    }
  }

  async cleanup() {
    for (const attachment of this.policyAttachmentsToSave) {
      await policyAttachmentStrategies[attachment.type].cleanup(attachment);
    }
  }
}

const policyAttachmentStrategies = {
  [PolicyType.opa]: {
    generate: async (input: PolicyInput) => {
      const path = await opaHelper.prepareCompiledRegoFile(input.metadata, input.code);
      return { path, metadata: input.metadata, type: PolicyType.opa };
    },
    cleanup: async (attachment: TempLocalPolicyAttachment) => {
      try {
        await opaHelper.deleteLocalRegoFile(attachment.path);
      } catch (err) {
        logger.warn(
          { err, attachment },
          'Failed cleanup of compiled rego file, this did not affect the request outcome'
        );
      }
    },
    saveToRepo: async (resourceRepository: ResourceRepository, attachment: TempLocalPolicyAttachment) => {
      const compiledRego = await opaHelper.readLocalRegoFile(attachment.path);
      const filename = opaHelper.getCompiledFilename(attachment.metadata);
      await resourceRepository.writePolicyAttachment(filename, compiledRego);
    },
    removeFromRepo: async (resourceRepository: ResourceRepository, metadata: ResourceMetadata) => {
      const filename = opaHelper.getCompiledFilename(metadata);
      await resourceRepository.deletePolicyAttachment(filename);
    },
  },
};
