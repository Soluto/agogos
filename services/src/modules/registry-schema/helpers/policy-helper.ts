import { parseType } from 'graphql';
import { PolicyInput } from '..';
import logger from '../../logger';
import { PolicyArgDefinition } from '../../resource-repository';

export function markOptionalPolicyArgs(policies?: PolicyInput[]) {
  policies?.forEach(p => {
    if (!p.args) return;

    Object.entries(p.args).forEach(([argName, argDef]: [string, PolicyArgDefinition]) => {
      try {
        const type = parseType(argDef.type);
        argDef.optional = type.kind !== 'NonNullType' || !!argDef.default;
      } catch (err) {
        logger.warn({ policy: p.metadata, argName, argDef, err }, 'Invalid policy argument');
        throw err;
      }
    });
  });
}
