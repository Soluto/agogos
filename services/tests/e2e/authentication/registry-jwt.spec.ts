import { gql, GraphQLClient } from 'graphql-request';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';
import { getToken, getInvalidToken } from '../../helpers/get-token';

const query = gql`
  query {
    plugins {
      name
      version
    }
  }
`;

describe('Authentication - Verify JWT', () => {
  const registryClient = new GraphQLClient('http://localhost:8090/graphql');

  beforeAll(() => {
    expect.addSnapshotSerializer(GraphQLErrorSerializer);
  });

  test('Valid JWT', async () => {
    const accessToken = await getToken({ scope: 'stitch-registry' });
    registryClient.setHeader('Authorization', `Bearer ${accessToken}`);

    const result = await registryClient.request(query);
    expect(result).toMatchSnapshot();
  });

  test('Invalid JWT', async () => {
    const accessToken = await getInvalidToken();
    registryClient.setHeader('Authorization', `Bearer ${accessToken}`);

    const result = await registryClient.request(query).catch(e => e.response);
    expect(result).toMatchSnapshot();
  });

  test('Invalid Authentication Scheme', async () => {
    const credentials = Buffer.from('user:pwd').toString('base64');
    registryClient.setHeader('Authorization', `Basic ${credentials}`);

    const result = await registryClient.request(query).catch(e => e.response);
    expect(result).toMatchSnapshot();
  });

  test('Unknown issuer', async () => {
    const accessToken = await getInvalidToken({ issuer: 'http://microsoft.com' });
    registryClient.setHeader('Authorization', `Bearer ${accessToken}`);

    const result = await registryClient.request(query).catch(e => e.response);
    expect(result).toMatchSnapshot();
  });

  test('Unexpected audience', async () => {
    const accessToken = await getInvalidToken({ audience: 'Stitch Other' });
    registryClient.setHeader('Authorization', `Bearer ${accessToken}`);

    const result = await registryClient.request(query).catch(e => e.response);
    expect(result).toMatchSnapshot();
  });

  test('JWK uri is not resolved', async () => {
    const accessToken = await getToken({ tokenEndpoint: 'http://localhost:8071/connect/token' });
    registryClient.setHeader('Authorization', `Bearer ${accessToken}`);

    const result = await registryClient.request(query).catch(e => e.response);
    expect(result).toMatchSnapshot();
  });

  test('Multiple audiences', async () => {
    const accessToken = await getToken({
      scope: 'stitch-registry stitch-gateway',
    });
    registryClient.setHeader('Authorization', `Bearer ${accessToken}`);

    const result = await registryClient.request(query);
    expect(result).toMatchSnapshot();
  });
});
