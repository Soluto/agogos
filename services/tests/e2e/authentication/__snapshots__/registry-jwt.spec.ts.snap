// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication - Verify JWT Invalid Authentication Scheme 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Unauthorized",
  "status": 401,
  "statusCode": 401,
}
`;

exports[`Authentication - Verify JWT Invalid JWT 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Unauthorized",
  "status": 401,
  "statusCode": 401,
}
`;

exports[`Authentication - Verify JWT JWK uri is not resolved 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Unauthorized",
  "status": 401,
  "statusCode": 401,
}
`;

exports[`Authentication - Verify JWT Multiple audiences 1`] = `
Object {
  "plugins": Array [
    Object {
      "name": "argument-injection-globals.js",
      "version": "N/A",
    },
    Object {
      "name": "async-function.js",
      "version": "N/A",
    },
    Object {
      "name": "crashing.js",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-index-file",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-node-module",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-npm-module",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-package-json-file",
      "version": "N/A",
    },
    Object {
      "name": "function.js",
      "version": "N/A",
    },
    Object {
      "name": "promise.js",
      "version": "N/A",
    },
    Object {
      "name": "single-file.js",
      "version": "N/A",
    },
    Object {
      "name": "transform-apollo-server-plugins",
      "version": "1.0.0",
    },
    Object {
      "name": "transform-base-schema",
      "version": "1.0.0",
    },
    Object {
      "name": "transform-resource-group.js",
      "version": "N/A",
    },
    Object {
      "name": "with-config.js",
      "version": "N/A",
    },
  ],
}
`;

exports[`Authentication - Verify JWT Unexpected audience 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Unauthorized",
  "status": 401,
  "statusCode": 401,
}
`;

exports[`Authentication - Verify JWT Unknown issuer 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Unauthorized",
  "status": 401,
  "statusCode": 401,
}
`;

exports[`Authentication - Verify JWT Valid JWT 1`] = `
Object {
  "plugins": Array [
    Object {
      "name": "argument-injection-globals.js",
      "version": "N/A",
    },
    Object {
      "name": "async-function.js",
      "version": "N/A",
    },
    Object {
      "name": "crashing.js",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-index-file",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-node-module",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-npm-module",
      "version": "N/A",
    },
    Object {
      "name": "folder-with-package-json-file",
      "version": "N/A",
    },
    Object {
      "name": "function.js",
      "version": "N/A",
    },
    Object {
      "name": "promise.js",
      "version": "N/A",
    },
    Object {
      "name": "single-file.js",
      "version": "N/A",
    },
    Object {
      "name": "transform-apollo-server-plugins",
      "version": "1.0.0",
    },
    Object {
      "name": "transform-base-schema",
      "version": "1.0.0",
    },
    Object {
      "name": "transform-resource-group.js",
      "version": "N/A",
    },
    Object {
      "name": "with-config.js",
      "version": "N/A",
    },
  ],
}
`;
